File: app/src/main/java/com/chuongvd/app/signal/data/source/local/datasource/SignalLocalDataSource.java
package com.chuongvd.app.signal.data.source.local.datasource;

import android.arch.lifecycle.LiveData;
import com.chuongvd.app.signal.data.database.AppDatabase;
import com.chuongvd.app.signal.data.entity.SignalEntity;
import com.chuongvd.app.signal.data.entity.UserEntity;
import java.util.List;

public class SignalLocalDataSource implements ISignalLocalDataSource {
    private static SignalLocalDataSource sInstance;

    private final AppDatabase mDatabase;

    private SignalLocalDataSource(AppDatabase database) {
        mDatabase = database;
    }

    public static SignalLocalDataSource getInstance(AppDatabase database) {
        if (sInstance == null) {
            synchronized (SignalLocalDataSource.class) {
                if (sInstance == null) {
                    sInstance = new SignalLocalDataSource(database);
                }
            }
        }
        return sInstance;
    }

    @Override
    public void saveSignal(SignalEntity signalEntity) {
        mDatabase.signalDao().insertSignal(signalEntity);
    }

    @Override
    public void saveListSignal(List<SignalEntity> list) {
        mDatabase.signalDao().updateSignal(list);
    }

    @Override
    public LiveData<List<SignalEntity>> getLocalSignals() {
        return mDatabase.signalDao().loadSignals();
    }
}

File: app/src/main/java/com/chuongvd/app/signal/data/source/local/datasource/SignalLocalDataSource.kt
package com.chuongvd.app.signal.data.source.local.datasource

import android.arch.lifecycle.LiveData
import com.chuongvd.app.signal.data.database.AppDatabase
import com.chuongvd.app.signal.data.entity.SignalEntity

class SignalLocalDataSource private constructor(
    private val mDatabase: AppDatabase) : ISignalLocalDataSource {

  override val localSignals: LiveData<List<SignalEntity>>
    get() = mDatabase.signalDao().loadSignals()

  override fun saveSignal(signalEntity: SignalEntity) {
    mDatabase.signalDao().insertSignal(signalEntity)
  }

  override fun saveListSignal(list: List<SignalEntity>) {
    mDatabase.signalDao().updateSignal(list)
  }

  companion object {
    private var sInstance: SignalLocalDataSource? = null

    fun getInstance(database: AppDatabase): SignalLocalDataSource {
      if (sInstance == null) {
        synchronized(SignalLocalDataSource::class.java) {
          if (sInstance == null) {
            sInstance = SignalLocalDataSource(database)
          }
        }
      }
      return this.sInstance!!
    }
  }
}
